Newsletters: add 288 (2024-02-07) (#1498)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2024-02-07-newsletter.md

name: 2024-02-07-newsletter
slug: 2024-02-07-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter FIXME

## News

- **Public disclosure of a block stalling bug in Bitcoin Core affecting LN:**
  Eugene Siegel [announced][siegel stall] to Delving Bitcoin a bug in
  Bitcoin Core he had [responsibly disclosed][topic responsible
  disclosures] almost three years ago.  Bitcoin Core 22 and higher
  contain fixes for the bug, but many people are still running affected
  versions and some of those users might also be running LN
  implementations or other contract protocol software that could be
  vulnerable to exploitation of the bug.  Upgrading to Bitcoin Core 22
  or higher is strongly recommended.  To the best of our knowledge, no
  one has lost funds due to the attack described below.

    An attacker finds an LN forwarding node that is associated with a
    relaying Bitcoin node running a version of Bitcoin Core earlier
    than 22.  The attacker opens many separate connections to a victim's
    Bitcoin node.  The attacker then attempts to deliver newly found
    blocks to the victim faster than any honest peers, resulting in

>     victim's node automatically assigning peers controlled by the

Reviewer: ```suggestion
    the victim's node automatically assigning peers controlled by the
```

    attacker to all of the victim's high-bandwidth [compact block
    relay][topic compact block relay] slots.

    After the attacker obtains control over many of the victim's Bitcoin
    peer slots, it uses channels it controls on either side of the
    victim to forward payments it creates.  For example:

      Attacker Spender -> Victim Forwarder -> Attacker Receiver

>     The attacker works with a miner create a block that unilaterally

Reviewer: ```suggestion
    The attacker works with a miner to create a block that unilaterally
```

    stalling attack to prevent the Bitcoin Core node from learning about
    the blocks containing the preimage.  The stalling attack takes
    advantage of older versions of Bitcoin Core being willing to wait up
    to 10 minutes for a peer to deliver a block it announced before
    requesting that block from another peer.  Given an average of 10
    minutes between blocks, that means an attacker who controls _x_
    connections can delay a Bitcoin node from receiving a block for
    roughly the time it takes to produce _x_ blocks.  If the forwarding
    payment has to be claimed within 40 blocks, an attacker controlling
    50 connections can have a reasonable chance of preventing the
    Bitcoin node from seeing the block containing the preimage until the
    spending node is able to receive a refund of the payment.  If that
    happens, the attacker's spending node paid nothing and the
    attacker's receiving node received an amount extracted from the
    victim's node.

    As Siegel reports, two changes were made to Bitcoin Core to prevent
    stalling:

    - [Bitcoin Core #22144][] randomizes the order in which peers are

>       serviced in the message handling thread.  See [Newsletter

Reviewer: ```suggestion
      serviced in the message-handling thread.  See [Newsletter
```

      #154][news154 stall].

    - [Bitcoin Core #22147][] keeps at least one outbound high bandwidth
      compact block peer even if inbound peers seem to be performing

>       better.  The local node selects its own outbound peers, meaning

Reviewer: ```suggestion
      better.  The local node selects its outbound peers, meaning
```

      they're less likely to be under the control of an attacker, so
      it's useful to keep at least one outbound peer for safety.

- **Securely opening zero-conf channels with v3 transactions:**
  Matt Corallo [posted][corallo 0conf] to Delving Bitcoin to discuss how
  to securely allow [zero-conf channel opening][topic zero-conf
  channels] when the proposed [v3 transaction relay policy][topic v3
  transaction relay] is being used.  Zero-conf channel opens are new
  single-funded channels where the funder gives some or all of their
  initial funds to the acceptor.  Those funds are not secure until the
  channel open transaction receives a sufficient number of
  confirmations, so there’s no risk to the acceptor spending some of
  those funds back through the funder using the standard LN protocol.
  The initial proposal for v3 transaction relay policy would only allow
  an unconfirmed v3 transaction to have, at most, a single child in the
  mempool; the expectation is that the single child will [CPFP fee
  bump][topic cpfp] its parent if necessary.

    Those v3 rules are incompatible with both parties being able to fee

>     bump a zero-conf channel open: the funding transaction which creates

Reviewer: ```suggestion
    bump a zero-conf channel open: the funding transaction that creates
```

    the channel is the parent of a v3 transaction which closes the
    channel and the grandparent of a v3 transaction for fee bumping.
    Since the v3 rules only allow one parent and one child, there's no

>     way for the funding transaction to be fee bumped without modifying

Reviewer: ```suggestion
    way for the funding transaction to be fee-bumped without modifying
```

    how it is created.  Bastien Teinturier [notes][teinturier splice]
    that [splicing][topic splicing] encounters a similar problem.

    As of this writing, the main solution proposed appeared to be
    modifying funding and splicing transactions to include an extra

>     output for CPFP fee bumping now, wait for [cluster mempool][topic

Reviewer: ```suggestion
    output for CPFP fee bumping now, waiting for [cluster mempool][topic
```

>     cluster mempool] to hopefully allow v3 to permit more permissive

Reviewer: ```suggestion
    cluster mempool] to allow v3 to permit more permissive
```

Author: Leaving as-is. I don't think we can say for certain that cluster mempool will make that possible.

Reviewer: That's fine, I think readers will understand what is meant, but I remember reading a long time ago that "hopefully" is often used incorrectly, and shouldn't be used at all in many cases. I can't recall the details, but I think the problem is (using this as an example), "hopefully" appears to be modifying "allow", which is not the intended meaning. To me, it reads as imprecise language (but maybe it's just me). Maybe something like:
```
As of this writing, the main solution proposed appeared to be
modifying funding and splicing transactions to include an extra
output for CPFP fee bumping now, in anticipation of cluster mempool
allowing v3 to permit more permissive topologies ...
```
(I think it's understood that anticipated events might never occur. -- especially in bitcoin!)

  create and sign a spend of funding transaction, which they keep
  offchain unless they need it.  After they've both signed the refund
  transaction, they can both safely sign and broadcast the parent
  funding transaction.  Because the child refund transaction depends on
  the parent funding transaction having the expected txid, this process
  is only safe if there's no risk of txid malleability.

    Segwit prevents txid malleability---but only if all inputs to the
    transaction spend segwit outputs from previous transactions.
    For segwit v0, the only way for Alice to be sure that Bob is
    spending a segwit v0 output is for her to obtain a copy of the
    entire previous transaction that contained Bob's output.  If Alice
    doesn't perform this check, Bob can lie about spending a segwit
    output and instead spend a legacy output that allows him to mutate
    the txid, allowing him to invalidate the refund transaction and refuse
    to return any funds to Alice unless she agrees to pay him a ransom.

    For segwit v1, each `SIGHASH_ALL` signature directly commits to
    every previous output being spent in the transaction (see

>     [Newsletter #97][news97 spk]), so Alice can require Bob disclose his

Reviewer: ```suggestion
    [Newsletter #97][news97 spk]), so Alice can require Bob to disclose his
```

    scriptPubKey (which she could learn anyway from other information
    Bob needs to disclose to create a shared transaction).  Alice
    verifies that scriptPubKey uses segwit, either v0 or v1, and has her
    signature commit to it.  Now, if Bob lied and actually had a
    non-segwit output, the commitment made by Alice's signature wouldn't
    be valid, so the signature wouldn't be valid, the funding
    transaction wouldn't confirm, and there would be no need for a
    refund.

    This leads to two rules that protocols depending on presigned
    refunds must follow for security:

>     1. If you are contributing a input, prefer to contribute an input

Reviewer: ```suggestion
    1. If you are contributing an input, prefer to contribute an input
```

  provide an update when the migration is complete.

## Releases and release candidates

*New releases and release candidates for popular Bitcoin infrastructure
projects.  Please consider upgrading to new releases or helping to test
release candidates.*

<!-- FIXME:harding to update Tuesday -->

## Notable code and documentation changes

_Notable recent changes in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], [Lightning BOLTs][bolts repo],
[Bitcoin Inquisition][bitcoin inquisition repo], and [BINANAs][binana
repo]._

- [Bitcoin Core #29189][] deprecates libconsensus.  Libconsensus was an
  attempt to make Bitcoin Core's consensus logic usable in other
  software.  However, the library hasn't seen any significant adoption
  and it has become a burden on maintenance of Bitcoin Core.  The plan
  is to "not migrate it to CMake and let it end with v27. Any remaining

>   use-cases could be handled in the future by libbitcoinkernel."

Reviewer 2: Perhaps libbitcoinkernel can link here: https://github.com/bitcoin/bitcoin/issues/27587

> - [Bitcoin Core #28956][] removes adjusted time from Bitcoin Core and

Reviewer 2: Could link adjusted time to https://bitcoinops.org/en/newsletters/2024/01/10/#bitcoin-core-pr-review-club

>   warns users if their computer's clock appears to out of sync with the

Reviewer 3: ```suggestion
  warns users if their computer's clock appears to be out of sync with the
```

  rest of the network.  Adjusted time was an automatic adjustment made
  to a local node's time based on the time reported by its peers.  This
  could help a node with a slightly incorrect clock to learn from its
  peers, allowing it to avoid unnecessarily rejecting blocks and also
  give blocks it produced a more accurate time.  However, adjusted time
  has also led to problems in the past and does not provide meaningful
  benefits to nodes on the current network.

- [Bitcoin Core #29347][] enables [v2 P2P transport][topic v2 p2p
  transport] by default.  New connections between two peers that both
  support the v2 protocol will use encryption.

> - [Core Lightning #6985][] adds options to `hsmtool` that allow it to

Reviewer: ```suggestion
- [Core Lightning #6985][] adds options to `hsmtool` that allows it to
```

Author: Leaving as-is.  I think the singular here agrees with the plural "options" earlier in the sentence.

Reviewer: The "it" here refers to hsmtool (singular), not options. But no big deal.

  return the private keys for the onchain wallet in a way that allows

>   those keys to imported into another wallet.

Reviewer: ```suggestion
  those keys to be imported into another wallet.
```

- [Core Lightning #6974][] makes various updates to CLN's connection
  and gossip management code.  A user-visible change is the addition of
  fields that indicate when a peer last had a stable connection to the
  local node for at least a minute.  This can allow removing peers with
  unstable connections.

- [Core Lightning #7022][] removes `lnprototest` from Core Lightning's
  testing infrastructure.  See [Newsletter #145][news145 lnproto] for
  our description of them being added.

- [Core Lightning #6936][] adds infrastructure to assist with

>   deprecating CLN features.  Features are now deprecated in code using a
>   functions that automatically disable those features by default based

Reviewer 3: ```suggestion
  deprecating CLN features.  Features are now deprecated in code using
  functions that automatically disable those features by default based
```

>   on the current program version.  Users can still force enable the

Reviewer 2: Maybe 
```suggestion
  on the current program version.  Users can still force-enable the
```

  features even after their indicated deprecation version as long as the
  code still exists.  This avoids an occasional problem where a CLN

>   feature would be reported as deprecated but still continue functioning

Reviewer: ```suggestion
  feature would be reported as deprecated but continues functioning
```

  by default for a long time after it was planned for removal, possibly
  leading users to continue depending on it and making actual removal
  more difficult.

- [LND #8345][] begins testing whether transactions are likely to relay
  before broadcasting them by calling a full node's `testmempoolaccept`
  RPC when available.  This allows the node to detect potential problems

>   with the transaction before anything is sent to a third-party,

Reviewer 3: ```suggestion
  with the transaction before anything is sent to a third party,
```

  bump][topic cpfp] its parent if necessary.

    Those v3 rules are incompatible with both parties being able to fee
    bump a zero-conf channel open: the funding transaction that creates
    the channel is the parent of a v3 transaction which closes the
    channel and the grandparent of a v3 transaction for fee bumping.
    Since the v3 rules only allow one parent and one child, there's no
    way for the funding transaction to be fee-bumped without modifying
    how it is created.  Bastien Teinturier [notes][teinturier splice]
    that [splicing][topic splicing] encounters a similar problem.

>     As of this writing, the main solution proposed appeared to be

Reviewer: ```suggestion
    As of this writing, the main solution proposed appears to be
```
